tests/GhcSession.hs

```diff
@@ -74,32 +76,49 @@ setup topdir act (srcdir, min_cabal_ver) = do
 
           setCurrentDirectory dir
           run "cabal" [ "configure" ]
---          run "cabal" [ "build" ]
 
           act dir
-  where
-    run x xs = do
-      print $ x:xs
-      o <- readProcess x xs ""
-      putStrLn o
-      return ()
+
+run :: String -> [String] -> IO ()
+run x xs = do
+  print $ x:xs
+  o <- readProcess x xs ""
+  putStrLn o
+  return ()
 
 test :: FilePath -> IO [Bool]
 test dir = do
     let qe = mkQueryEnv dir (dir </> "dist")
-    cs <- runQuery qe $ components $ (,,) <$> entrypoints <.> ghcOptions
-    forM cs $ \(ep, opts, cn) -> do
-        let opts' = "-Werror" : opts
-        let sopts = intercalate " " $ map formatArg $ "ghc" : opts'
+    let packageDir = dir </> "dist" </> "package.conf.inplace"
+    cs <- runQuery qe $ components $ (,,,) <$> entrypoints <.> ghcOptions <.> needsBuildOutput
+    forM cs $ \(ep, opts, nb, cn) -> do
+
+        putStrLn $ "\n" ++ show cn ++ ":::: " ++ show nb
+
+        when (nb == ProduceBuildOutput) $ do
+          run "cabal" [ "build" ]
+
+        exists <- doesDirectoryExist packageDir
+        let opts' = if exists
+              then ("-package-db " ++ packageDir) : "-Werror" : opts
+              else "-Werror" : opts
```

Comment:
This whole `packageDir` business shouldn absolutely not be here, if it has to be something in Main.hs is still broken. Remember this test is supposed to simulate what ghc-mod will have to do and the whole point of cabal-helper is to abstract away cabal specific hacks like that.

Can you explain why this is necessary?

Reply:
It is a matter of timing.

If a component has `ProduceBuildOutput` set, then invoking `cabal build` (at the ghc-mod or higher level) will result in the `packageDir` being created.

But at the point we invoke cabal-helper the build has not yet happened.

So one alternative is to invoke cabal-helper more than once, if `ProduceBuildOutput` is set, and let it check for existence of that package db dir and add it to the options if needed.

But that does seem more complicated than requiring the cabal-helper client to do it.

@DanielG please give your preference.